openmc/mesh.py

```diff
@@ -2673,74 +2674,53 @@ def write_data_to_vtk(
         Parameters
         ----------
         filename : str or pathlib.Path
-            Name of the VTK file to write. If the filename ends in '.vtu' then a
-            binary VTU format file will be written, if the filename ends in
-            '.vtk' then a legacy VTK file will be written.
+            Name of the VTK file to write. If the filename ends in '.vtkhdf'
+            then a VTKHDF format file will be written. If the filename ends in
+            '.vtu' then a binary VTU format file will be written, if the
+            filename ends in '.vtk' then a legacy VTK file will be written.
         datasets : dict
             Dictionary whose keys are the data labels and values are numpy
             appropriately sized arrays of the data
         volume_normalization : bool
             Whether or not to normalize the data by the volume of the mesh
             elements
         """
-        from vtkmodules.util import numpy_support
-        from vtkmodules import vtkCommonCore
-        from vtkmodules import vtkCommonDataModel
-        from vtkmodules import vtkIOLegacy
-        from vtkmodules import vtkIOXML
-
-        if self.connectivity is None or self.vertices is None:
-            raise RuntimeError(
-                "This mesh has not been loaded from a statepoint file."
-            )
-
-        if filename is None:
-            filename = f"mesh_{self.id}.vtk"
-
-        if Path(filename).suffix == ".vtk":
-            writer = vtkIOLegacy.vtkUnstructuredGridWriter()
-
-        elif Path(filename).suffix == ".vtu":
-            writer = vtkIOXML.vtkXMLUnstructuredGridWriter()
-            writer.SetCompressorTypeToZLib()
-            writer.SetDataModeToBinary()
-
-        writer.SetFileName(str(filename))
 
-        grid = vtkCommonDataModel.vtkUnstructuredGrid()
-
-        points = vtkCommonCore.vtkPoints()
-        points.SetData(numpy_support.numpy_to_vtk(self.vertices))
-        grid.SetPoints(points)
-
-        n_skipped = 0
-        for elem_type, conn in zip(self.element_types, self.connectivity):
-            if elem_type == self._LINEAR_TET:
-                elem = vtkCommonDataModel.vtkTetra()
-            elif elem_type == self._LINEAR_HEX:
-                elem = vtkCommonDataModel.vtkHexahedron()
-            elif elem_type == self._UNSUPPORTED_ELEM:
-                n_skipped += 1
-                continue
-            else:
-                raise RuntimeError(f"Invalid element type {elem_type} found")
-
-            for i, c in enumerate(conn):
-                if c == -1:
-                    break
-                elem.GetPointIds().SetId(i, c)
-
-            grid.InsertNextCell(elem.GetCellType(), elem.GetPointIds())
-
-        if n_skipped > 0:
-            warnings.warn(
-                f"{n_skipped} elements were not written because "
-                "they are not of type linear tet/hex"
+        if Path(filename).suffix == ".vtkhdf":
```

Comment:
These different paths for IO have made this method pretty large. Can you look into refactoring the different output paths into other internal methods?